core/src/test/java/uk/ac/bbsrc/tgac/miso/core/data/workflow/impl/TestWorkflowTest.java

```diff
@@ -0,0 +1,255 @@
+    assertEquals(Collections.emptyList(), workflow.getLog());
+  }
+
+  private <T extends Throwable> void assertThrows(Class<T> expectedType, Runnable runnable) {
+    try {
+      runnable.run();
+    } catch (Throwable actualException) {
+      if (!expectedType.isInstance(actualException)) {
+        throw new AssertionFailedError(String.format("Expected exception %s was not thrown", expectedType.toString()));
+      }
+    }
+  }
+
+  @Test
+  public void testSetProgressTwiceThrowsError() {
+    workflow = new TestWorkflow();
+    workflow.setProgress(makeProgress());
+
+    assertThrows(IllegalStateException.class, () -> workflow.setProgress(makeProgress()));
+  }
+
+  @Test
+  public void testCreateNewWorkflowWithIncorrectWorkflowName() {
+    workflow = new TestWorkflow();
+    assertThrows(IllegalArgumentException.class, () -> workflow.setProgress(makeProgress(LOADSEQUENCER)));
+  }
+
+  @Test
+  public void testProcessInvalidInput() {
+    workflow = makeNewWorkflow();
+    assertThrows(IllegalArgumentException.class, () -> workflow.processInput(makePoolStep(POOL_ID)));
+  }
+
+  @Test
+  public void testProcessValidInput() {
+    workflow = makeNewWorkflow();
+    workflow.processInput(makeIntegerStep(INT_1));
+    assertReceivedOneInput(workflow, INT_1);
+  }
+
+  private void assertReceivedOneInput(Workflow workflow, int input) {
+    assertThrows(IllegalArgumentException.class, () -> workflow.processInput(2, makePoolStep(POOL_ID)));
+
+    assertEquivalent(makeProgress(makeIntegerStep(input, 0)), workflow.getProgress());
+    assertPoolPrompt(workflow.getNextStep());
+    assertIntegerPrompt(workflow.getStep(0));
+    assertPoolPrompt(workflow.getStep(1));
+    assertThrows(IllegalArgumentException.class, () -> workflow.getStep(2));
+    assertFalse(workflow.isComplete());
+    assertEquals(Collections.singletonList(String.format("Processed integer %d", input)), workflow.getLog());
+  }
+
+  @Test
+  public void testProcessInputAtFirstStep() {
+    workflow = makeNewWorkflow();
+    workflow.processInput(0, makeIntegerStep(INT_1));
+    assertReceivedOneInput(workflow, INT_1);
+  }
+
+  @Test
+  public void testCancelInputWithoutInput() {
+    makeNewWorkflow().cancelInput();
+  }
+
+  @Test
+  public void loadExistingWorkflow() {
+    assertReceivedOneInput(makeExistingWorkflow(makeIntegerStep(INT_1)), INT_1);
+  }
+
+  private Workflow makeExistingWorkflow(ProgressStep... steps) {
+    Workflow workflow = new TestWorkflow();
+    workflow.setProgress(makeProgress(steps));
+    return workflow;
+  }
+
+  @Test
+  public void testProcessInputAfterLoadingInput() {
+    workflow = makeExistingWorkflow(makeIntegerStep(INT_1));
+    workflow.processInput(makePoolStep(POOL_ID));
+    assertReceivedTwoInputs(workflow, INT_1, POOL_ID);
+  }
+
+  private void assertReceivedTwoInputs(Workflow workflow, int input1, long input2Id) {
+    assertThrows(IllegalArgumentException.class, () -> workflow.processInput(2, makePoolStep(POOL_ID)));
```

Comment:
It should be possible to change input for any step, even after it is "completed" as long as the workflow hasn't been executed yet

Reply:
Good point.
In this case, the step with index 2 is actually the "3rd" step which does not exist in the 2-step `TestWorkflow`, so I think this behaviour should be invalid.  However, I have (perhaps excessively) stringent validation of the `stepNumber` parameter to both `getStep` and `processInput`.  I can remove these assertions and validations to communicate the future `Workflow`s can be flexible about how they handle processing input out of order.